envoy/tracing/trace_context.h

```diff
@@ -60,11 +97,11 @@ class TraceContext {
    * @param key The context key of string view type.
    * @param val The context value of string view type.
    */
-  virtual void setTraceContextReference(absl::string_view key, absl::string_view val) {
+  virtual void setByReference(absl::string_view key, absl::string_view val) {
```

Comment:
By default, the behaviors of `setByReference` and `setByReferenceKey` and `setByKey` are exactly the same, and derived classes can ignore them. But `setByReference` and `setByReferenceKey` have additional constraints on the life cycle of key/value. Derived classes that wish to use reference semantics to improve performance or reduce memory overhead can override these methods.

Taking the current HTTP implementation as an example, when setByReferenceKey is called, only the value will be copied. 

I am not sure if this can prove the value of these two methods. Or is the overhead of a small amount of memory copy acceptable? 

```cpp
void RequestHeaderMapImpl::setByKey(absl::string_view key, absl::string_view val) {
  ASSERT(validatedLowerCaseString(key));
  HeaderMapImpl::removeExisting(key);

  HeaderString new_key;
  new_key.setCopy(key);
  HeaderString new_val;
  new_val.setCopy(val);

  HeaderMapImpl::insertByKey(std::move(new_key), std::move(new_val));
}

void RequestHeaderMapImpl::setByReferenceKey(absl::string_view key, absl::string_view val) {
  ASSERT(validatedLowerCaseString(key));
  HeaderMapImpl::removeExisting(key);

  HeaderString new_val;
  new_val.setCopy(val);

  HeaderMapImpl::insertByKey(HeaderString(key), std::move(new_val));
}
```

Reply:
Oh I see. This is to map to the header map implementation. I think it's strange to provide default versions on the interface then. Can you make these pure virtual?

Reply:
This is of course no problem.